core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -916,7 +807,232 @@ object Chunk {
+        case _           => Arr(self.toArray(Chunk.classTagOf(self)))
+      }
+
+    final def mapAccum[S1, B](s1: S1)(f1: (S1, A) => (S1, B)): (S1, NonEmpty[B]) = {
+      val len            = self.length
+      val init           = f1(s1, head)
+      val dest: Array[B] = Array.ofDim(len)(Chunk.Tags.fromValue(init._2))
+
+      dest(0) = init._2
+      var i     = 1
+      var s: S1 = init._1
+
+      while (i < len) {
+        val t = f1(s, self(i))
+
+        s = t._1
+        dest(i) = t._2
+
+        i += 1
+      }
+
+      (s, Chunk.Arr(dest))
+    }
+
+    final def mapM[R, E, B](f: A => ZIO[R, E, B]): ZIO[R, E, NonEmpty[B]] = {
+      val len = self.length
+
+      val init: ZIO[R, E, B] = f(head)
+
+      var array: ZIO[R, E, Array[B]] = init.map { b =>
+        implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+        val array                   = Array.ofDim[B](len)
+        array(0) = b
+        array
+      }
+
+      var i = 1
+
+      while (i < len) {
+        val j = i
+        array = array.zipWith(f(self(j))) { (array, b) =>
+          array(j) = b
+          array
+        }
+
+        i += 1
+      }
+
+      array.map(array => Chunk.Arr(array))
+    }
+
+    override final def mapMPar[R, E, B](f: A => ZIO[R, E, B]): ZIO[R, E, NonEmpty[B]] = {
+      val len = self.length
+
+      val init: ZIO[R, E, B] = f(head)
+
+      var array: ZIO[R, E, Array[B]] = init.map { b =>
+        implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+        val array                   = Array.ofDim[B](len)
+        array(0) = b
+        array
+      }
+
+      var i = 1
+
+      while (i < len) {
+        val j = i
+        array = array.zipWithPar(f(self(j))) { (array, b) =>
+          array(j) = b
+          array
+        }
+
+        i += 1
+      }
+
+      array.map(array => Chunk.Arr(array))
+    }
+
+    /**
+     * Statefully and effectfully maps over the elements of this chunk to produce
+     * new elements.
+     */
+    final def mapAccumM[R, E, S1, B](s1: S1)(f1: (S1, A) => ZIO[R, E, (S1, B)]): ZIO[R, E, (S1, NonEmpty[B])] = {
+
+      val len = self.length
+
+      val init: ZIO[R, E, (S1, B)] = f1(s1, head)
+
+      var dest: ZIO[R, E, (S1, Array[B])] = init.map {
+        case (s, b) =>
+          implicit val B: ClassTag[B] = Chunk.Tags.fromValue(b)
+          val array                   = Array.ofDim[B](len)
+          array(0) = b
+          (s, array)
+      }
+
+      var i = 1
+
+      while (i < len) {
+        val j = i
+        dest = dest.flatMap {
+          case (state, array) =>
+            f1(state, self(j)).map {
+              case (state2, b) =>
+                array(j) = b
+                (state2, array)
+            }
+        }
+
+        i += 1
+      }
+
+      dest.map {
+        case (state, array) => (state, Chunk.Arr(array))
+      }
+    }
+
+    final def ++[A1 >: A](that: Chunk[A1]): NonEmpty[A1] = Chunk.concat(self, that)
+
+    final def head: A        = self(0)
```

Comment:
`head` / `tail` really only make sense with linked list structures.

Personally, I'd rename `head` => `first`, and then omit `tail`. Why omit tail? Because it's not a performant operation. If you recursively "uncons" from a chunk, you'll build up n-levels of allocation and indirection by the end—severe performance penalty.